john: 1.9.0-Jumbo-1 -> rolling-2404

CherryKitten opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Nixpkgs is currently still on a 5 year old release, because upstream has not been tagging new stable releases for a very long time.
They do provide some form of rolling release in a seperate repository (https://github.com/openwall/john-packages/releases/tag/rolling-2404), so I have decided to use the same git commit hash as was used in the latest rolling release.

Some parts of the software have been broken on nixpkgs for a while. Updating to this rolling release will for example fix https://github.com/NixOS/nixpkgs/issues/267978, which has been open for almost a year now.

I've also added myself to the maintainer list.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

itslychee commented:
looks like there's a new release being planned here as well!

https://github.com/openwall/john/pull/5494

litchipi commented:
When trying to use this version to crack a `yescrypt` hash, it tells me he cannot find any `john.conf` in the nix store `/share/john` directory.
`fopen: /nix/store/slkgvc2kkpiiz5pgyjjgzgpgqka70y2w-john-rolling-2404/share/john/john.conf: No such file or directory`

And when I setup a local `--config=john.conf` replacement, it seeks for a `dumb16.conf` which he doesn't find either.
I'm not really familiar with how john works internally, but it seams some files are missing from the install phases ?

CherryKitten commented:
> looks like there's a new release being planned here as well!
> 
> [openwall/john#5494](https://github.com/openwall/john/pull/5494)

huh, created a day after my PR here, looks like the old issue being pushed up again might have set something in motion.

@litchipi i'll have a closer look soon, john is a pretty big and complex project so there is a good chance I might have missed parts in the update process that are only neccessary for use-cases I didn't test myself

CherryKitten commented:
> When trying to use this version to crack a `yescrypt` hash, it tells me he cannot find any `john.conf` in the nix store `/share/john` directory. `fopen: /nix/store/slkgvc2kkpiiz5pgyjjgzgpgqka70y2w-john-rolling-2404/share/john/john.conf: No such file or directory`
> 
> And when I setup a local `--config=john.conf` replacement, it seeks for a `dumb16.conf` which he doesn't find either. I'm not really familiar with how john works internally, but it seams some files are missing from the install phases ?

this should be fixed now, this was covered by the steps in the `postPath` phase, which I had removed together with the patches, despite still being needed.

I added them back and tried it out, it no longer asks for the config file

litchipi commented:
Ok, it fixed the config problem error, but I still have a `yescrypt` problem.
Could you test it on your machine ? Running john on this hash `root:$y$j9T$AjVXCCcjJ6jTodR8BwlPf.$4NeBwxOq4X0/0nCh3nrIBmwEEHJ6/kDU45031VFCWc2` (it comes from HackMyVM)

It seams that even with `libxcrypt` present (at least I see it listed in `ldd`), it doesn't detect that the system supports `yescrypt`.
Browsed a bit the sources, I thought it could come from https://github.com/openwall/john/blob/6be5461b67951290f707021b93b34ea4168fa9e2/src/c3_fmt.c#L341, OpenMP calling the `crypt_r` function, which apparently is at the collision between `unistd` and `libxcrypt` (as explained here: https://github.com/besser82/libxcrypt/blob/72f75aa370ae96ccd2cc44ea3cf4182d8679ffbe/lib/crypt-port.h#L53)
But maybe I got it all wrong, not really sure what I'm reading tbh.

Tell me if that's the for another issue, or if we can try to tackle this here before merging the update ?

CherryKitten commented:
> Ok, it fixed the config problem error, but I still have a `yescrypt` problem. Could you test it on your machine ? Running john on this hash `root:$y$j9T$AjVXCCcjJ6jTodR8BwlPf.$4NeBwxOq4X0/0nCh3nrIBmwEEHJ6/kDU45031VFCWc2` (it comes from HackMyVM)
> 
> It seams that even with `libxcrypt` present (at least I see it listed in `ldd`), it doesn't detect that the system supports `yescrypt`. Browsed a bit the sources, I thought it could come from https://github.com/openwall/john/blob/6be5461b67951290f707021b93b34ea4168fa9e2/src/c3_fmt.c#L341, OpenMP calling the `crypt_r` function, which apparently is at the collision between `unistd` and `libxcrypt` (as explained here: https://github.com/besser82/libxcrypt/blob/72f75aa370ae96ccd2cc44ea3cf4182d8679ffbe/lib/crypt-port.h#L53) But maybe I got it all wrong, not really sure what I'm reading tbh.
> 
> Tell me if that's the for another issue, or if we can try to tackle this here before merging the update ?

yeah, i'm getting the same problem and also can't figure out a solution right now, sorry :/

Maybe since this was broken before already, it'd make sense to merge the update, since that fixes various other issues, and then keep focussing on the yescrypt problem seperately in a new issue

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4113

JohnRTitor commented:
Result of `nixpkgs-review pr 318620` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>adenum</li>
    <li>john</li>
    <li>johnny</li>
    <li>wifite2</li>
    <li>wifite2.dist</li>
  </ul>
</details>

JohnRTitor reviewed (approved):
Builds and runs, lgtm

JohnRTitor commented on pkgs/tools/security/john/default.nix:
```diff
@@ -1,32 +1,17 @@
 stdenv.mkDerivation rec {
   pname = "john";
-  version = "1.9.0-jumbo-1";
+  version = "rolling-2404";
```
Unstable versioning should be formatted like: `1.9.0-jumbo-unstable-YYYY-MM-DD`

CherryKitten replied:
are you sure? "rolling-2024" is a version name upstream uses in their unstable packages, so i went with that over a custom nixpkgs specific version name

JohnRTitor replied:
According #234201, we should include the preceeding version if we are packaging something untagged. `rolling-2024` does not really a "version tag", is it?

CherryKitten replied:
it is a version tag upstream at https://github.com/openwall/john-packages/releases/tag/rolling-2404

i don't have a strong opinion on it either way, i can still change it when I'm able to tomorrow if you think it's really important

JohnRTitor replied:
Oh and looks like previous versions are also tagged https://github.com/openwall/john-packages/releases/tag/rolling-2310 https://github.com/openwall/john-packages/releases/tag/rolling-2304 

Then I am open to keeping the current version `rolling-2404`. And instead of the commit, the tag can be directly fetched I suppose? See my comment below

CherryKitten replied:
that doesn't work, because the tag is in a seperate repo, upstream is a bit messy there..

they've tagged that version in their "john-packages" repo, which contains some binary builds and instructions for packaging and stuff, but it's not tagged in the actual "john" repo containing the code of the project, which is the one we fetch

that's why I've just used the same commit hash that's referenced in their package repo

CherryKitten replied:
to make this a bit clearer

the upstream organization has two repos

john
john-packages

john has the actual code and some old version tags, john-packages has some scripts for packaging 

on john-packages, there is an official version tag for rolling-2404, with packaging instructions that refer back to the john repo at a specific commit hash

that is why i used the version name from john-packages and directly fetch the corresponding code from john

JohnRTitor replied:
Generic update scripts/bots are gonna have a hard time dealing with that. Wish there was a update script. 

If you'd like to design an update script for it, you can fetch the tag from `john-packages` repo, and somehow get the commit hash from `Created-on_2024-04-03.txt` release file. Then substitute the commit hash in the package definition. 

I'd assume that would be a lot complex to design, I am not opposed to the idea of fetching the latest commit/making it stay on bleeding edge all the time. There is `unstableGitUpdater` for that too, so bots like r-ryantm can easily update it even in a maintainer crisis.

CherryKitten replied:
This package has already not been updated in like 5 years, I'd rather just get any more recent version merged than bikeshedding about update-scripts to be honest..

I fixed the other comments you had, but I would prefer to just stay on the version and method we have and not do any more big changes, something like update scripts might be a thing for a future follow-up in my opinion

JohnRTitor commented on pkgs/tools/security/john/default.nix:
```diff
@@ -1,32 +1,17 @@
 
   src = fetchFromGitHub {
     owner = "openwall";
     repo = pname;
```
```suggestion
    repo = "john";
```

JohnRTitor reviewed (commented):
Could you address these before I merge?

JohnRTitor commented on pkgs/tools/security/john/default.nix:
```diff
@@ -1,32 +1,17 @@
     repo = pname;
-    rev = "1.9.0-Jumbo-1";
-    sha256 = "sha256-O1iPh5QTMjZ78sKvGbvSpaHFbBuVc1z49UKTbMa24Rs=";
+    rev = "f9fedd238b0b1d69181c1fef033b85c787e96e57";
```
Then can we use 
```suggestion
    rev = "ref/tags/${version}";
```

JohnRTitor replied:
Not applicable

JohnRTitor commented on pkgs/tools/security/john/default.nix:
```diff
@@ -1,32 +1,17 @@
-    rev = "1.9.0-Jumbo-1";
-    sha256 = "sha256-O1iPh5QTMjZ78sKvGbvSpaHFbBuVc1z49UKTbMa24Rs=";
+    rev = "f9fedd238b0b1d69181c1fef033b85c787e96e57";
+    sha256 = "sha256-zvoN+8Sx6qpVg2JeRLOIH1ehfl3tFTv7r5wQZ44Qsbc=";
```
`hash` is preferred

```suggestion
    hash = "sha256-zvoN+8Sx6qpVg2JeRLOIH1ehfl3tFTv7r5wQZ44Qsbc=";
```

CherryKitten committed:
john: 1.9.0-Jumbo-1 -> rolling-2404

CherryKitten committed:
john: add cherrykitten as maintainer
